packages/tailwindcss/src/apply.ts

```diff
@@ -176,7 +176,54 @@ export function substituteAtApply(ast: AstNode[], designSystem: DesignSystem) {
         let candidates = Object.keys(candidateOffsets)
         let compiled = compileCandidates(candidates, designSystem, {
           onInvalidCandidate: (candidate) => {
-            throw new Error(`Cannot apply unknown utility class: ${candidate}`)
+            // When using prefix, make sure prefix is used in candidate
+            if (designSystem.theme.prefix && !candidate.startsWith(designSystem.theme.prefix)) {
+              throw new Error(
+                `Cannot apply unprefixed utility class: \`${candidate}\`, did you mean \`${designSystem.theme.prefix}:${candidate}\`?`,
+              )
+            }
+
+            // When the utility is blocklisted, let the user know
+            //
+            // Note: `@apply` is processed before handling incoming classes from
+            // template files. This means that the `invalidCandidates` set will
+            // only contain explicit classes via:
+            //
+            // - `blocklist` from a JS config
+            // - `@source not inline(…)`
+            if (designSystem.invalidCandidates.has(candidate)) {
+              throw new Error(
+                `Cannot apply \`${candidate}\`, it seems like the utility was explicitly excluded and cannot be applied.\n\nMore info: https://tailwindcss.com/docs/detecting-classes-in-source-files#explicitly-excluding-classes`,
+              )
+            }
+
+            // Verify `@tailwind utilities` or `@reference` is used
+            let hasUtilitiesOrReference = false
+            walk(ast, (node, { context }) => {
+              // Find `@reference`
+              if (context.reference) {
+                hasUtilitiesOrReference = true
+                return WalkAction.Stop
+              }
+
+              // Find `@tailwind utilities`
+              else if (
+                node.kind === 'at-rule' &&
+                node.name === '@tailwind' &&
+                (node.params === 'utilities' || node.params.startsWith('utilities'))
```

Comment:
Let's check the `designSystem` itself so it:

- works when there's only a theme file inlined
- we can check it in constant time
- it will also work if the theme config is only coming from legacy plugins